modules/33acrossAnalyticsAdapter.js

```diff
@@ -0,0 +1,461 @@
+ * @property {AdUnitSize} size
+ */
+
+/**
+ * @typedef {Object} Bid - Parsed bid data
+ * @property {string} bidder
+ * @property {string} bidId
+ * @property {string} source
+ * @property {string} status
+ * @property {BidResponse} [bidResponse]
+ * @property {1|0} [hasWon]
+ */
+
+/**
+ * @typedef {Object} AdUnit - Parsed adUnit data
+ * @property {string} transactionId - Primary key for *this* auction/adUnit combination
+ * @property {string} adUnitCode
+ * @property {string} slotId - Equivalent to GPID. (Note that
+ * GPID supports adUnits where multiple units have the same `code` values
+ * by appending a `#UNIQUIFIER`. The value of the UNIQUIFIER is likely to be the div-id,
+ * but, if div-id is randomized / unavailable, may be something else like the media size)
+ * @property {Array<AdUnitMediaType>} mediaTypes
+ * @property {Array<AdUnitSize>} sizes
+ * @property {Array<Bid>} bids
+ */
+
+/**
+ * After the first bid is initiated, we wait until every bid is completed before sending the report.
+ *
+ * We will listen for the `bidWon` event to determine when all bids are complete.
+ */
+class TransactionManager {
+  #timeoutId;
+  #pending = 0;
+  #timeout;
+  #transactions = {};
+  #onComplete;
+
+  get #unsent() {
+    return this.#pending;
+  }
+
+  set #unsent(value) {
+    this.#pending = value;
+
+    if (this.#pending <= 0) {
+      this.#clearTimeout();
+      this.#onComplete();
+      this.#transactions = {};
+    }
+  }
+
+  constructor({ timeout, onComplete }) {
+    this.#timeout = timeout;
+    this.#onComplete = onComplete;
+  }
+
+  initiate(transactionId) {
+    if (this.#transactions[transactionId]) return;
+
+    this.#transactions[transactionId] = {
+      status: 'waiting'
+    };
+    ++this.#unsent;
+
+    this.#restartSendTimeout();
+  }
+
+  complete(transactionId) {
+    if (!this.#transactions[transactionId]) {
+      log.warn(`transactionId "${transactionId}" was not found. Nothing to enqueue.`);
+      return;
+    }
+    this.#transactions[transactionId].status = 'queued';
+    --this.#unsent;
+
+    log.info(`Queued transaction "${transactionId}" for delivery. ${this.#unsent} transactions still pending.`, this.#transactions);
+  }
+
+  completeAll(reason) {
+    Object.keys(this.#transactions).forEach(transactionId => {
+      this.complete(transactionId);
+    });
+
+    log.info('All remaining transactions flushed.' + (reason ? ` Reason: ${reason}` : ''));
+  }
+
+  // gulp-eslint is using eslint 6, a version that doesn't support private method syntax
+  // eslint-disable-next-line no-dupe-class-members
+  #clearTimeout() {
+    return window.clearTimeout(this.#timeoutId);
+  }
+
+  // eslint-disable-next-line no-dupe-class-members
+  #restartSendTimeout() {
+    this.#clearTimeout();
+
+    this.#timeoutId = setTimeout(() => {
+      if (this.#timeout !== 0) {
+        log.warn(`Timed out waiting for ad transactions to complete. Sending report.`);
+      }
+
+      this.#unsent = 0;
```

Comment:
Here it would be nice to call the flushTransactions() method I mentioned above so that your intentions are a bit more clear. You might also need to add a line this.#pending = 0 before that to keep state consistent for now, but I'm going to address this.#pending separately as I'm not sure this counter is needed.